Remora.Discord.Commands/Parsers/MessageParser.cs

```diff
@@ -0,0 +1,71 @@
+//
+//  MessageParser.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) 2017 Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System.Threading;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Remora.Commands.Parsers;
+using Remora.Commands.Results;
+using Remora.Discord.API;
+using Remora.Discord.API.Abstractions.Objects;
+using Remora.Discord.API.Abstractions.Rest;
+using Remora.Discord.Commands.Contexts;
+using Remora.Discord.Commands.Extensions;
+using Remora.Results;
+
+namespace Remora.Discord.Commands.Parsers;
+
+/// <summary>
+/// Parses instances of <see cref="IMessage"/> from command-line inputs.
+/// </summary>
+[PublicAPI]
+public class MessageParser : AbstractTypeParser<IMessage>
+{
+    private readonly ICommandContext _context;
+    private readonly IDiscordRestChannelAPI _channelAPI;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="MessageParser"/> class.
+    /// </summary>
+    /// <param name="context">The command context.</param>
+    /// <param name="channelAPI">The channel API.</param>
+    public MessageParser(ICommandContext context, IDiscordRestChannelAPI channelAPI)
+    {
+        _channelAPI = channelAPI;
+        _context = context;
+    }
+
+    /// <inheritdoc />
+    public override async ValueTask<Result<IMessage>> TryParseAsync
+    (
+        string value,
+        CancellationToken ct = default
+    )
+    {
+        if (!DiscordSnowflake.TryParse(value.Unmention(), out var messageID))
```

Comment:
It's probably also useful to parse message links here, since those allow us to look up messages in channels other than the current context's channel.

Reply:
By message link I suppose you mean these links?

In Guild: 
https://discord.com/channels/{GUILD_ID}/{CHANNEL_ID}/{MESSAGE_ID}
https://canary.discord.com/channels/{GUILD_ID}/{CHANNEL_ID}/{MESSAGE_ID}

In DM:
https://discord.com/channels/@me/{CHANNEL_ID}/{MESSAGE_ID}
https://canary.discord.com/channels/@me/{CHANNEL_ID}/{MESSAGE_ID}

I would actually implement the check for link using RegEx:
`https:\/\/(canary\.)?discord\.com\/channels\/(?<guild_id>@me|[0-9]*)\/(?<channel_id>[0-9]*)\/(?<message_id>[0-9]*)$`

Reply:
Yep, that's correct - I'd use a compiled regex as well.

Reply:
regex source generator to the rescue here then.

Reply:
ptb.discord.com also needs to be matched

Reply:
Added handling of message links in `MessageParser`